### ISSUE TITLE
Need binaries for up to date FreeBSD 13.

### ISSUE THREAD

**Issue by arrowd** [Thefrank/freebsd-dotnet-sourcebuild#2]
FreeBSD 12 is going EoL in December so the least support FreeBSD version would be 13.2
The problem with FreeBSD 12 binaries is that they are linked to OpenSSL 1.1, while FreeBSD 13 ships OpenSSL 1.1.1
This makes it impossible to use FreeBSD 12 binaries as bootstrap on FreeBSD 13 (on 14 too, but that's for later)

Can you please provide binaries that are compiled on the fully updated FreeBSD 13.2?

**Comment by Thefrank** [comment]
Quick run test in VM:
```
root@freebsd:~ # uname -a; freebsd-version -kru
FreeBSD freebsd 13.2-RELEASE-p4 FreeBSD 13.2-RELEASE-p4 GENERIC amd64
13.2-RELEASE-p4
13.2-RELEASE-p4
13.2-RELEASE-p4
root@freebsd:~ # fetch https://github.com/Thefrank/freebsd-dotnet-sourcebuild/releases/download/v8.0.100/dotnet-sdk-8.0.100-rc.2.23502.1-freebsd-x64.tar.gz
root@freebsd:~ # mkdir dotnetrc2
root@freebsd:~ # tar xf dotnet-sdk-8.0.100-rc.2.23502.1-freebsd-x64.tar.gz -C dotnetrc2/
root@freebsd:~ # elfctl -e +noaslr dotnetrc2/dotnet   
root@freebsd:~ # dotnetrc2/dotnet --info
.NET SDK:
 Version:   8.0.100-rc.2.23502.1
 Commit:    0abacfc2b6

Runtime Environment:
 OS Name:     FreeBSD
 OS Version:  13
 OS Platform: FreeBSD
 RID:         freebsd-x64
 Base Path:   /root/dotnetrc2/sdk/8.0.100-rc.2.23502.1/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.0-rc.2.23475.17
  Architecture: x64
  Commit:       0b25e38ad3

.NET SDKs installed:
  8.0.100-rc.2.23502.1 [/root/dotnetrc2/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.0-rc.2.23480.2 [/root/dotnetrc2/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.0-rc.2.23475.17 [/root/dotnetrc2/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
root@freebsd:~ # ldd dotnetrc2/shared/Microsoft.NETCore.App/8.0.0-rc.2.23475.17/libSystem.Security.Cryptography.Native.OpenSsl.so
dotnetrc2/shared/Microsoft.NETCore.App/8.0.0-rc.2.23475.17/libSystem.Security.Cryptography.Native.OpenSsl.so:
        libcrypto.so.11 => /usr/local/lib/libcrypto.so.11 (0x6b4f1ef4000)
        libssl.so.11 => /usr/local/lib/libssl.so.11 (0x6b4f19e1000)
        libc.so.7 => /lib/libc.so.7 (0x6b4ee39a000)
        libthr.so.3 => /lib/libthr.so.3 (0x6b4f3051000)
```
(I did not post the `pkg install` segment)
Can you post the error you are getting?

**Comment by arrowd** [comment]
I'm running `FreeBSD 13.2-RELEASE-p5` in the jail and it doesn't have `/lib/libcrypto.so.11`, but `/lib/libcrypto.so.111`.

How do you update your installation? `freebsd-update` or from source?

**Comment by Thefrank** [comment]
I generally try and use the VM images provided by FreeBSD.

Please don't build against base OpenSSL when using the VMR. `runtime` will not even look at `/lib/` as it sets `-DFEATURE_DISTRO_AGNOSTIC_SSL=0`
You CAN build against base OpenSSL when using each repo on its own as that defaults to `-DFEATURE_DISTRO_AGNOSTIC_SSL=1` and will use `libdl` to access OpenSSL.

See: https://github.com/dotnet/runtime/issues/14537#issuecomment-1822781974 for a more complete explanation.

**Comment by arrowd** [comment]
I still don't get it. `dlopen` doesn't ignore `/lib/` so it still should use OpenSSL from base. The `FEATURE_DISTRO_AGNOSTIC_SSL` flag just removes the .so dependency from the ELF file.

Anyways, using OpenSSL from ports is a no-go for porting.

**Comment by Thefrank** [comment]
I guess I do not understand what you are asking? 
Take two SDKs: one build either via crossbuilding or native and one built using the VMR.
Do this:

`find /PATH/TO/DOTNET/ -name "*OpenSsl*.so" -exec ldd "{}" ";"` 
What do each link against? The VMR should link against: libcrypto.so.XX, libssl.so.XX, libc.so.7, and libthr.so.3. 
The crossbuilt/nativebuilt should link against libdl.so, libthr.so.3, and libc.so.7.

from my port:
```
root@freebsd14:/usr/ports/lang/dotNET8 # find work/stage -name "*OpenSsl*.so" -exec ldd "{}" ";"
work/stage/usr/local/dotnet/shared/Microsoft.NETCore.App/8.0.0/libSystem.Security.Cryptography.Native.OpenSsl.so:
        libcrypto.so.12 => /usr/local/lib/libcrypto.so.12 (0x125c774eb000)
        libssl.so.12 => /usr/local/lib/libssl.so.12 (0x125c79022000)
        libc.so.7 => /lib/libc.so.7 (0x125c76d5f000)
        libthr.so.3 => /lib/libthr.so.3 (0x125c79997000)
work/stage/usr/local/dotnet/packs/Microsoft.NETCore.App.Runtime.freebsd-x64/8.0.0/runtimes/freebsd-x64/native/libSystem.Security.Cryptography.Native.OpenSsl.so:
        libcrypto.so.12 => /usr/local/lib/libcrypto.so.12 (0x69e33e2e000)
        libssl.so.12 => /usr/local/lib/libssl.so.12 (0x69e3272c000)
        libc.so.7 => /lib/libc.so.7 (0x69e303d1000)
        libthr.so.3 => /lib/libthr.so.3 (0x69e33303000)
```
(from https://github.com/sec/dotnet-core-freebsd-source-build/releases/tag/8.0.100-x64-native):
```
root@freebsd14:~ # find dn/ -name "*OpenSsl*.so" -exec ldd "{}" ";"
dn/shared/Microsoft.NETCore.App/8.0.0/libSystem.Security.Cryptography.Native.OpenSsl.so:
        libdl.so.1 => /usr/lib/libdl.so.1 (0x261dddc4000)
        libthr.so.3 => /lib/libthr.so.3 (0x261dec90000)
        libc.so.7 => /lib/libc.so.7 (0x261dd763000)
```

The VMR is not designed to generate portable SDKs. It is designed for linux distros that demand sourcebuilt packages. It works for FreeBSD's port system because any step after `fetch` is removed the network and the VMR can be built entirely offline after `fetch`.

> Anyways, using OpenSSL from ports is a no-go for porting.

If true `rm -r /usr/ports/lang/dotNET8/` and I finally get to experience "finished product not launched due to external factors"

I also want my weekend back. I couldve been sleeping or being lazy during that time!

**Comment by arrowd** [comment]
> If true rm -r /usr/ports/lang/dotNET8/ and I finally get to experience "finished product not launched due to external factors"

Why? What prevents using base OpenSSL? What happen if you do `pkg remove -x openssl` and then try to build VMR?

**Comment by Thefrank** [comment]
> Why? What prevents using base OpenSSL? What happen if you do `pkg remove -x openssl` and then try to build VMR?

If using a non-portable bootstrap. dotNET fails to start because of the missing libraries.
If using a portable bootstrap, runtime fails to build as it can not find a working OpenSSL via `cmake`'s `find_package()`

Atleast on FreeBSD 14, there are some oddities between base and ports OpenSSL that prevent simple libmap overrides.
```
readelf -a /lib/libcrypto.so.30 | grep OpenSSL_version
  1040: 0000000000199ff0    11 FUNC    GLOBAL DEFAULT   14 OpenSSL_version_num@@OPENSSL_1_1_0 (2)
  4292: 000000000019a050   139 FUNC    GLOBAL DEFAULT   14 OpenSSL_version@@OPENSSL_1_1_0 (2)
readelf -a /usr/local/lib/libcrypto.so | grep OpenSSL_version
  2130: 00000000002ea7f0    11 FUNC    GLOBAL DEFAULT   14 OpenSSL_version_num@@OPENSSL_3.0.0 (2)
  3501: 00000000002ea850   139 FUNC    GLOBAL DEFAULT   14 OpenSSL_version@@OPENSSL_3.0.0 (2)
```

**Comment by arrowd** [comment]
> If using a portable bootstrap, runtime fails to build as it can not find a working OpenSSL via cmake's find_package()

Then let's just fix that?

**Comment by Thefrank** [comment]
> Then let's just fix that?

Yep! I have tried a number of solutions including `PATHS`, `HINTS`, `min`, and `OPENSSL 3` but all failed. I am absolutely open to suggestions on this.

**Comment by arrowd** [comment]
If we're talking about the `FindOpenSSL.cmake` bundled with CMake itself then according to the documentation, passing `-DOPENSSL_ROOT_DIR=/usr` somewhere to CMake should be enough. Of course, the `openssl` should not be installed on the system.

If you tried that, do you still have error logs by hand?

**Comment by sec** [comment]
Sorry for my 2 cents, but, when I did VMR build under FreeBSD 14:
- took SDK build under 13.2, not from VMR, but from each repo, it wasn't linked with any openssl, but using libdl as @Thefrank wrote already
- did VMR build
- result SDK is linked with base system OpenSSL (cut from log: Found OpenSSL: /usr/lib/libcrypto.so (found version "3.0.12"))
- and in fact it is - libcrypto.so.30 => /lib/libcrypto.so.30 (0x26cde56e5000) - now this SDK works without having openssl from packages installed

So, question, @arrowd what SDK do you need?

If you need SDK for 13.2 - you need OpenSSL3 which needs to be installed from ports/packages, as base doesn't have 3.
For 14.0 - SDK build under 13.2 will need OpenSSL from packages - but I have build of VMR done under VMR, which doesn't require openssl from ports/packages, becuase it's linked with the system one.

**Comment by arrowd** [comment]
> If you need SDK for 13.2 - you need OpenSSL3 which needs to be installed from ports/packages, as base doesn't have 3.

And .NET doesn't build against 1.1? Why does it have to be 3?

**Comment by sec** [comment]
I think it should build - crossbuilds/native ones are using libdl so they should work with 1.1 (just checked on 13.2-RELEASE with base openssl and https://github.com/sec/dotnet-core-freebsd-source-build/releases/download/8.0.100-x64-native/dotnet-sdk-8.0.100-freebsd-x64.tar.gz) - looks like it's working.

edit: keep in mind that 1.1.1 is EOL and 14 is using 3 - for me, it makes more sense to have 13+openssl3 and 14+base rather that fighting with 1.1.1.

**Comment by arrowd** [comment]
The problem is that we can't have 13+openssl3. The openssl port is sort of special in that it can't be just installed into the system by some dependency. Instead it requires an user to set `DEFAULT_VERSION=ssl=openssl3` and then build all ports from source. This will make sure that all ports software is linked to the same openssl library.

Anyways, I'll now proceed with current SDK for 13.2 and try to come up with steps for recreating the same SDK without dependency on the ports openssl. Let's see how this will go down.

**Comment by Thefrank** [comment]
github needs multiquote...

> took SDK build under 13.2, not from VMR, but from each repo, it wasn't linked with any openssl, but using libdl as @Thefrank wrote already
>did VMR build

@sec which prebuilt binaries tarball did you use? I am trying to figure out where this hard dependency comes in.

>Anyways, I'll now proceed with current SDK for 13.2 and try to come up with steps for recreating the same SDK without dependency on the ports openssl. Let's see how this will go down.

@arrowd `portlint` does not warn against having `openssl` in any of the depends lines which is why I assumed it was OK. Only `stage-qa` gives warnings, not errors, about lack of USES=ssl or DEFAULT_VERSION=ssl=opensslX. 


I am going to step away from this for a few days. If I can't figure out the libssl depends issue when I get back...

"`BROKEN_FreeBSD_13` in Makefile and hope that is OK for ports@ `||` `rm -r /usr/ports/lang/dotNET8/` to avoid duplicate work"

**Comment by arrowd** [comment]
Yes, we can always just give up on FreeBSD 13.

**Comment by sec** [comment]
Why drop 13? To build VMR under 13, all is needed is SDK bootstraped NOT from VMR (ex. either crossbuild or native build from each repo)
Building by hand from each repo make the build with FEATURE_DISTRO_AGNOSTIC_SSL=1
Building from VMR makes FEATURE_DISTRO_AGNOSTIC_SSL=0

And that flag is controller by PortableBuild, right?
Or the problem is that, build pickup openssl3 (when there's 1 and 3 both available) ?

For 14 we can use SDK from VMR (as it's linked to base openssl3, which will work, but it's hacky I think)

**Comment by Thefrank** [comment]
welp. this is the best that I can do: Consumes portable RC2. No hard depend on OpenSSL. `USES=ssl`. Outputs non-portable SDK that uses base OpenSSL3. Tested for FreeBSD 14 only. Need to wire in `Flavors` for FreeBSD 13 support

---

SDK FROM: https://github.com/sec/dotnet-core-freebsd-source-build/releases/download/8.0.100-vmr/dotnet-sdk-8.0.100-freebsd-x64.tar.gz
```
root@freebsd14:~/dn # find ./ -name "*Ssl*.so" -exec readelf -a "{}" ";" | grep _version
00000001fcc0 000c00000007 R_X86_64_JUMP_SLOT  0000000000000000 X509_get_version + 0
00000001fe28 003900000007 R_X86_64_JUMP_SLOT  0000000000000000 OpenSSL_version_num + 0
0000000205b0 012400000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_get_version + 0
0000000205e0 012a00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_version + 0
000000020708 014f00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_CIPHER_get_version + 0
    12: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version@OPENSSL_3.0.0 (3)
    57: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num@OPENSSL_3.0.0 (3)
   292: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version@OPENSSL_3.0.0 (4)
   298: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version@OPENSSL_3.0.0 (4)
   335: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version@OPENSSL_3.0.0 (4)
    53: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS _version.c
    79: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version
   147: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num
   563: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version
   575: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version
   640: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version
  0x0000 vn_version: 1 vn_file: libcrypto.so.12 vn_cnt: 1
  0x0010 vn_version: 1 vn_file: libssl.so.12 vn_cnt: 1
  0x0020 vn_version: 1 vn_file: libc.so.7 vn_cnt: 3
00000001fcc0 000c00000007 R_X86_64_JUMP_SLOT  0000000000000000 X509_get_version + 0
00000001fe28 003900000007 R_X86_64_JUMP_SLOT  0000000000000000 OpenSSL_version_num + 0
0000000205b0 012400000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_get_version + 0
0000000205e0 012a00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_version + 0
000000020708 014f00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_CIPHER_get_version + 0
    12: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version@OPENSSL_3.0.0 (3)
    57: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num@OPENSSL_3.0.0 (3)
   292: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version@OPENSSL_3.0.0 (4)
   298: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version@OPENSSL_3.0.0 (4)
   335: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version@OPENSSL_3.0.0 (4)
    53: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS _version.c
    79: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version
   147: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num
   563: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version
   575: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version
   640: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version
  0x0000 vn_version: 1 vn_file: libcrypto.so.12 vn_cnt: 1
  0x0010 vn_version: 1 vn_file: libssl.so.12 vn_cnt: 1
  0x0020 vn_version: 1 vn_file: libc.so.7 vn_cnt: 3
```
VMR PORTS:
```
root@freebsd14:/usr/ports/lang/dotNET8/work/stage/usr/local/dotnet # find ./ -name "*Ssl*.so" -exec readelf -a "{}" ";" | grep _version
00000001fe00 000c00000007 R_X86_64_JUMP_SLOT  0000000000000000 X509_get_version + 0
00000001ff68 003900000007 R_X86_64_JUMP_SLOT  0000000000000000 OpenSSL_version_num + 0
0000000206f0 012400000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_get_version + 0
000000020720 012a00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_version + 0
000000020848 014f00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_CIPHER_get_version + 0
    12: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version@OPENSSL_1_1_0 (4)
    57: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num@OPENSSL_1_1_0 (4)
   292: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version@OPENSSL_1_1_0 (5)
   298: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version@OPENSSL_1_1_0 (5)
   335: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version@OPENSSL_1_1_0 (5)
  0x0000 vn_version: 1 vn_file: libcrypto.so.30 vn_cnt: 3
  0x0010 vn_version: 1 vn_file: libssl.so.30 vn_cnt: 3
  0x0020 vn_version: 1 vn_file: libc.so.7 vn_cnt: 3
00000001fe00 000c00000007 R_X86_64_JUMP_SLOT  0000000000000000 X509_get_version + 0
00000001ff68 003900000007 R_X86_64_JUMP_SLOT  0000000000000000 OpenSSL_version_num + 0
0000000206f0 012400000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_get_version + 0
000000020720 012a00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_version + 0
000000020848 014f00000007 R_X86_64_JUMP_SLOT  0000000000000000 SSL_CIPHER_get_version + 0
    12: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND X509_get_version@OPENSSL_1_1_0 (4)
    57: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND OpenSSL_version_num@OPENSSL_1_1_0 (4)
   292: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_get_version@OPENSSL_1_1_0 (5)
   298: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_version@OPENSSL_1_1_0 (5)
   335: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND SSL_CIPHER_get_version@OPENSSL_1_1_0 (5)
  0x0000 vn_version: 1 vn_file: libcrypto.so.30 vn_cnt: 3
  0x0010 vn_version: 1 vn_file: libssl.so.30 vn_cnt: 3
  0x0020 vn_version: 1 vn_file: libc.so.7 vn_cnt: 3
```
---
@arrowd Have more ports related questions. Base provides `libunwind`. runtime provides in-tree `libunwind`. FreeBSD's version rejects both and uses ports. Is this a show stopper? Same question for GSSAPI. Ports version is used.

It was my understanding that items in base are for general usage and building/using the OS. Stuff in ports is for ports. So having `DEPENDS` would better than `USES`. If this policy has changed then a large number of changes need to be made in how runtime consumes and uses various libraries under FreeBSD

@sec the main flag to stop the build system of dotnet/* from using online resource is `/p:DotNetBuildFromSource`. This turns off PGO (not used by FreeBSD), NativeAOT, Agnostic SSL, and others that I can not think of right now. This also stops most, but not all, repos from generating a portable RID.

**Comment by Thefrank** [comment]
Added WIP for ports

**Comment by arrowd** [comment]
I came up with a port based on @Thefrank 's version. It builds fine on 15-CURRENT, although there might be a problem with aslr.

Anyways, now my questions are:

1. Can we generate `bootstrapsdk` distfiles from the built port? If so, how?
2. Same question for `bootstrapartifacts` distfiles.
3. What is the `Private.SourceBuilt.Artifacts.${DOTNET_BOOTSTRAPARTIFACTS}.centos.8-x64.tar.gz` distfile? I'm confused by the `centos` part. Should we generate it for FreeBSD? Again, is is possible to do from the built port's `WRKSRC`?
4. What versions `8.0.0` and `8.0.100` are correctly called? Is the former "dotnet version" and the latter - "sdk version"?

Thanks again both you guys for all the effort you put into this. We're really close into getting the port into the main tree!

**Comment by Thefrank** [comment]
@arrowd 

1+2: you want SHA512's for all of the files contained in both of those packages? It is doable but I don't have a place to host that many files and I assumed (incorrectly??) that size and SHA512 matching two large files would be sufficient protection against bad actors.  Also, those files are generated during the final step of the source build to be used as input for the next version of dotNET (e.g., NET8RC2 generates the 8.0.100 SDK for users and an artifacts ball to be used with the SDK for building 8.0.1/8.0.xxx from source)

From what I recall source building does not provide deterministic compilation. If you are looking for way to verify via signature the source code: https://github.com/dotnet/dotnet/releases each release has a .sig with it.

3: Easiest way to get all of the required items during fetch. 98% of the files not platform specific and the build infrastructure does not care if there are extra items just items that are missing. Most are generated from https://github.com/dotnet/source-build-reference-packages

4: 8.0.0 is for items that don't need SDK versioning. 8.0.100 is for things that do. "Runtimes" (e.g., netcore and aspnetcore) use the former and SDKs/Installers use the later. Example in production: https://github.com/dotnet/core/releases/tag/v8.0.0 .Different repos use different tags for the results "core" product

`dotnet` allows for many different SDKs and runtimes to live side-by-side
```
frank@buildbot:~$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.417
 Commit:    04fde1d2e3

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.417/

global.json file:
  Not found

Host:
  Version:      6.0.25
  Architecture: x64
  Commit:       492abbeef0

.NET SDKs installed:
  5.0.408 [/usr/share/dotnet/sdk]
  6.0.417 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.25 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.25 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Download .NET:
  https://aka.ms/dotnet-download

Learn about .NET Runtimes and SDKs:
  https://aka.ms/dotnet/runtimes-sdk-info
```
```
[root@azurecli ~]# dotnet --info
.NET SDK:
 Version:   7.0.203
 Commit:    5b005c19f5

Runtime Environment:
 OS Name:     FreeBSD
 OS Version:  13
 OS Platform: FreeBSD
 RID:         freebsd.13-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.203/

Host:
  Version:      7.0.5
  Architecture: x64
  Commit:       8042d61b17

.NET SDKs installed:
  6.0.302 [/usr/share/dotnet/sdk]
  6.0.401 [/usr/share/dotnet/sdk]
  6.0.404 [/usr/share/dotnet/sdk]
  6.0.406 [/usr/share/dotnet/sdk]
  6.0.408 [/usr/share/dotnet/sdk]
  7.0.202 [/usr/share/dotnet/sdk]
  7.0.203 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.7 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.9 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.14 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.16 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.9 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.14 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.16 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
[root@azurecli ~]# 
```

**Comment by arrowd** [comment]
> 1+2: you want SHA512's for all of the files contained in both of those packages?

No, I need a way to regenerate, say, the `runtime.freebsd-x64.Microsoft.NETCore.ILAsm.nupkg` file. After getting the port installed, what commands do I need to generate it?

**Comment by Thefrank** [comment]
from the nuspec:
```xml
<description>Internal implementation package not meant for direct consumption. Please do not reference directly. 
The .NET IL Assembler.</description>
```
It is in `Private.SourceBuilt.Artifacts.DOTNETVERSION.OS-ARCH.tar.gz` if you absolutely need it. The SDK should not nor should end users of the SDK need it.
If you want to pull out individual packages from the source build remove `--clean-while-building` and fish them out at the end. Between "Shipping" and "Nonshipping" there should be ~500 nuget packages, most of which were restored and not produced.

**Comment by arrowd** [comment]
I have minified the port to not have any unnecessary patches, distfiles and steps. The port from https://github.com/arrowd/freebsd-ports/tree/dotnet is now ready to be upstreamed. Please take a look at it and leave comments if any.

I'll now contact FreeBSD `portmgr` team asking to enable `allow.mlock=true` for the build jails on our jailer. This is the last problem holding us off from getting the port into the tree.

Meanwhile, @sec @Thefrank @wfurt share with me your real names and email addresses if you want me to give you a credit in the commit message. It is OK to omit the real name and use a nickname instead. If you know someone else who also contributed to our cause, please tag them too.

And the last question - anyone wants to take a maintainership of the newly created port? I will still help with carrying out the builds, pushing the stuff into the tree and port-related fixes, but I'm lacking the domain knowledge of .NET itself. You mined a lot of domain knowledge while hacking on it, so it'd be great if someone steps out for the maintarnership.

Again, thanks everyone involved! This is a tremendous success and I'm happy it got to the fruition.

**Comment by Thefrank** [comment]
Ignore-able to nitpick:

---
nitpick

https://github.com/arrowd/freebsd-ports/blob/f9d9c5adbb1b9eb024886d55240613b1e13517e3/lang/dotnet/files/patch-src_installer_src_redist_targets_GenerateBundledVersions.targets

The patch I made here adds `freebsd-x64` RID support for FreeBSD builds that, while available on the internet, are not `pkg install`able. Not a problem per se but it will create confusion for anyone trying to target an older version dotNET only to discover they need to wander the internet to find the needed NuGet files to make the build happen. Users should not be targeting EOL versions of dotNET but I left it this way so older code could be maintained. The impact here is close to, if not, 0 users.

---
ignore-able

https://github.com/Thefrank/freebsd-dotnet-sourcebuild/blob/a4eca45cfbb3eed5dddf2e352264feaf534721ed/PortWIP/dotNET8/files/patch-src_runtime_src_native_libs_System.Security.Cryptography.Native_opensslshim.c

This patch updates the helper function for finding usable versions of OpenSSL. Useful for building portable builds where newer versioned OpenSSL exists or non-portables where the input and output will have different OpenSSL versions. This can be forced in cases where the system is using a version of OpenSSL that is not on the list. Verify the output library is linked correctly and that it looks for the correct symbols. If the build runs without it and without the override on all in-support FreeBSD versions then this patch is unneeded. 

---
ignore-able

`eng/bootstrap/OverrideBootstrapVersions.props` 

My `Makefile` generates this to copy one of the functions that `prep.sh` does: fixing earlier build authoring/versioning errors. Unnecessary if the build completes without it. Future versions are unlikely to need this.

---

Thanks for getting this over the finishing line!

**Comment by arrowd** [comment]
Do you think, we should remove [patch-src_installer_src_redist_targets_GenerateBundledVersions.targets](https://github.com/arrowd/freebsd-ports/blob/f9d9c5adbb1b9eb024886d55240613b1e13517e3/lang/dotnet/files/patch-src_installer_src_redist_targets_GenerateBundledVersions.targets) then?

> If the build runs without it and without the override on all in-support FreeBSD versions then this patch is unneeded.

Yep, it is, which is why I removed it. The only little problem is that bootstrap binaries are linked to `111`, but it is easily fixable by creating a `misc/compat13x` port. I already mailed maintainer of `misc/compat12x` asking him if he can do this.

> eng/bootstrap/OverrideBootstrapVersions.props

Yep, I also removed that change.

**Comment by sec** [comment]
@arrowd have you tested the building the port under 14? For me it's failing due to missing openssl when using bootstrapped SDK it downloaded `Shared object "libcrypto.so.111" not found, required by "libSystem.Security.Cryptography.Native.OpenSsl.so"`

I assume creating `misc/compat13x` will fix this by installing previous openssl version?

**Comment by arrowd** [comment]
> I assume creating misc/compat13x will fix this by installing previous openssl version?

Exactly. I already contacted `compat12x` maintainer asking for this. If I don't get an answer in near days, I'll create this port myself and this will unblock building on FreeBSD 14+

**Comment by Thefrank** [comment]
Using a portable build of dotNET as a bootstrap should (hopefully) resolve the issue without needing to generate more dependencies/ports for it.

**Comment by arrowd** [comment]
I don't quite understand how could it be truly portable. It'd still need to use some version of OpenSSL?

**Comment by Thefrank** [comment]
> I don't quite understand how could it be truly portable. It'd still need to use some version of OpenSSL?

dotNET is not capable of building true "portable" binaries for any given OS. It can however work with a large number of OpenSSL versions, `libunwind` types, node versions, and *libc libraries if generated as a "portable" version.

The portable version uses the helper function and links to `ldl` to find a usable version of OpenSSL as long as that version ends in .10, .11, or .111, it will work. You can override this check and force it to look for a version of OpenSSL with a certain number. 

https://github.com/Thefrank/freebsd-dotnet-sourcebuild/blob/a4eca45cfbb3eed5dddf2e352264feaf534721ed/PortWIP/dotNET8/Makefile#L45

The helper function is patched as more of a "we should never get here" worst case of the VMR not linking to a version of OpenSSL but instead `ldl`. VMR on Linux can take either a portable build or a non-portable distro-specific build and output a new non-portable distro-specific build. This was my failed attempt at replicating that.

Feel free to ignore any/all that I have done with the Makefile. I don't mind being wrong; I just want it working :)

It has been over 7 years since the first people starting work FreeBSD support for dotNET. I am just happy to see it finally finish out as a port.

**Comment by arrowd** [comment]
> It can however work with a large number of OpenSSL versions, libunwind types, node versions, and *libc libraries if generated as a "portable" version.

Ah, I see. Anyways, I don't think it is worth bothering to build bootstrap in this mode. It is much easier to provide missing bits as `BUILD_DEPENDS`.

**Comment by Thefrank** [comment]
@arrowd how long will port files be available on distcache.freebsd.org/local-distfiles/arrowd/dotnet/ ? Am I safe to delete my artifacts here and/or delete the repo?

**Comment by arrowd** [comment]
> @arrowd how long will port files be available on distcache.freebsd.org/local-distfiles/arrowd/dotnet/ ?

Until I remove them.

> Am I safe to delete my artifacts here and/or delete the repo?

Yes, you are, but why do that? They might still be useful for someone and the whole repo may serve historic purposes.

**Comment by Thefrank** [comment]
> Yes, you are, but why do that? They might still be useful for someone and the whole repo may serve historic purposes.

This has only happened to me on a private repo, but at some point GH gets annoyed about how much space your repo's releases take up. 
With an official port. this repo and a few others of mine would be at-best redundant to the port or at-worst confusing to those looking for dotNET on FreeBSD. 

I did just discover that GH allows you to archive a repo! I think this would be better way of guiding people towards an official port rather than a repo disappearing.

**Comment by Thefrank** [comment]
Closing. We have a version in ports!

Thank you @arrowd for your work here!
